File: python/leetcode/editor/cn/P104_MaximumDepthOfBinaryTree.py
from collections import deque
from typing import Optional

from leetcode.editor.cn.dataStruct.TreeNode import TreeNode, createTree


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    depth = 0

    def maxDepth1(self, root: Optional[TreeNode]) -> int:
        def func(node: TreeNode) -> int:
            if node is None:
                return 0
            left = func(node.left) + 1
            right = func(node.right) + 1
            return max(left, right)

        return func(node=root)

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        res = 0
        cur = deque([root])
        while cur:
            res += 1
            for _ in range(len(cur)):
                node = cur.pop()
                if node.left:
                    cur.appendleft(node.left)
                if node.right:
                    cur.appendleft(node.right)
        return res


# leetcode submit region end(Prohibit modification and deletion)
print(Solution().maxDepth(createTree([3, 9, 20, None, None, 15, 7])))

File: java/leetcode/src/leetcode/editor/cn/P104_MaximumDepthOfBinaryTree.java
//给定一个二叉树，找出其最大深度。 
//
// 二叉树的深度为根节点到最远叶子节点的最长路径上的节点数。 
//
// 说明: 叶子节点是指没有子节点的节点。 
//
// 示例： 
//给定二叉树 [3,9,20,null,null,15,7]， 
//
//     3
//   / \
//  9  20
//    /  \
//   15   7 
//
// 返回它的最大深度 3 。 
// Related Topics 树 深度优先搜索 广度优先搜索 二叉树 👍 1279 👎 0

package leetcode.editor.cn;


import leetcode.editor.cn.DataStruct.TreeNode;

import java.util.ArrayDeque;

//Java：二叉树的最大深度
class P104_MaximumDepthOfBinaryTree {
    public static void main(String[] args) {
        Solution solution = new P104_MaximumDepthOfBinaryTree().new Solution();
    }
    //leetcode submit region begin(Prohibit modification and deletion)

    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     * int val;
     * TreeNode left;
     * TreeNode right;
     * TreeNode() {}
     * TreeNode(int val) { this.val = val; }
     * TreeNode(int val, TreeNode left, TreeNode right) {
     * this.val = val;
     * this.left = left;
     * this.right = right;
     * }
     * }
     */
    class Solution {
        public int maxDepth(TreeNode root) {
            if (root == null) {
                return 0;
            }
            int depth = 0;
            ArrayDeque<TreeNode> arrayDeque = new ArrayDeque<>();
            arrayDeque.offer(root);
            while (!arrayDeque.isEmpty()) {
                int len = arrayDeque.size();
                while (len > 0) {
                    TreeNode p = arrayDeque.poll();
                    if (p.left != null) {
                        arrayDeque.offer(p.left);
                    }
                    if (p.right != null) {
                        arrayDeque.offer(p.right);
                    }
                    --len;
                }
                ++depth;
            }
            return depth;
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}
